subprojects/persistent-cache/src/integTest/groovy/org/gradle/cache/internal/DefaultFileLockManagerContentionIntegrationTest.groovy

```diff
@@ -0,0 +1,318 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.cache.internal
+
+import org.gradle.cache.FileLock
+import org.gradle.cache.FileLockManager
+import org.gradle.cache.internal.filelock.LockOptionsBuilder
+import org.gradle.cache.internal.locklistener.DefaultFileLockContentionHandler
+import org.gradle.cache.internal.locklistener.FileLockContentionHandler
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.executer.GradleHandle
+import org.gradle.internal.concurrent.DefaultExecutorFactory
+import org.gradle.internal.remote.internal.inet.InetAddressFactory
+import org.gradle.internal.time.TrueTimeProvider
+
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.TimeoutException
+
+import static org.gradle.test.fixtures.ConcurrentTestUtil.poll
+
+class DefaultFileLockManagerContentionIntegrationTest extends AbstractIntegrationSpec {
+    def timeProvider = new TrueTimeProvider()
+    def addressFactory = new InetAddressFactory()
+
+    FileLockContentionHandler receivingFileLockContentionHandler
+    DatagramSocket receivingSocket
+    FileLock receivingLock
+    Thread socketReceiverThread
+
+    def setup() {
+        executer.withArguments("-d")
+        executer.requireOwnGradleUserHomeDir().withDaemonBaseDir(file("daemonsRequestingLock")).requireDaemon()
+        buildFile << ""
+    }
+
+    def cleanup() {
+        socketReceiverThread?.terminate = true
+    }
+
+    def "the lock holder is not hammered with ping requests for the shared fileHashes lock"() {
+        given:
+        setupLockOwner()
+        def pingRequestCount = 0
+        //do not handle requests: this simulates the situation were pings do not arrive
+        replaceSocketReceiver { pingRequestCount++ }
+
+        when:
+        def build = executer.withTasks("help").start()
+        def startTime = timeProvider.currentTime
+        poll {
+            assert (build.standardOutput =~ 'Pinged owner at port').count == 3
+        }
+        receivingLock.close()
+
+        then:
+        build.waitForFinish()
+        pingRequestCount == 3 || pingRequestCount == 4
+        timeProvider.currentTime - startTime > 3000 // See: DefaultFileLockContentionHandler.PING_DELAY
+    }
+
+    def "the lock holder starts the release request only once and discards additional requests in the meantime"() {
+        given:
+        def requestReceived = false
+        setupLockOwner {
+            requestReceived = true
+            while(true) { Thread.sleep(100000) } //block the release action thread
```

Comment:
When the test worker stops :)? Good catch. Added a clean up to the end of the test.